### ISSUE TITLE
setuptools 60: "No module named 'setuptools'" when installing Poetry from source

### ISSUE THREAD

**Issue by alkasm** [pypa/setuptools#2980]
### setuptools version

60.0.0

### Python version

3.9.6

### OS

macOS 11.5.2

### Description

Installing a package which uses a Poetry backend from source fails on 60.0.0. Poetry packages that I've tried (my own, as well as Poetry itself) fail to install. Package installs are successful after reverting to previous setuptools versions.

The [changelog](https://github.com/pypa/setuptools/blob/main/CHANGES.rst) states that v60.0.0 introduces a breaking change:
> pypa/setuptools#2896: Setuptools once again makes its local copy of distutils the default. To override, set SETUPTOOLS_USE_DISTUTILS=stdlib.

When using the environment variable as suggested, the package installs are again successful.

I tried a few flit and setuptools projects and observed no issues with those.

### Expected behavior

The package should install without error.

### How to Reproduce

Poetry itself uses poetry as a build tool, so installing poetry from source fails. 

```
pip install --upgrade pip setuptools==60.0.0
unset SETUPTOOLS_USE_DISTUTILS
pip install git+ssh://git@github.com/python-poetry/poetry
```

Alternatively, clone poetry and run `pip install .` locally inside the repo.

### Output

<details>
<summary>pip install git+ssh://git@github.com/python-poetry/poetry</summary>

```
pip install git+ssh://git@github.com/python-poetry/poetry
Collecting git+ssh://****@github.com/python-poetry/poetry
  Cloning ssh://****@github.com/python-poetry/poetry to /private/var/folders/k6/2dzjkwsd1mj8m52xz1b2swym0000gp/T/pip-req-build-oq1rbxm7
  Running command git clone --filter=blob:none -q 'ssh://****@github.com/python-poetry/poetry' /private/var/folders/k6/2dzjkwsd1mj8m52xz1b2swym0000gp/T/pip-req-build-oq1rbxm7
  Resolved ssh://****@github.com/python-poetry/poetry to commit 6a35cc41e950d4ff0f1efad649c1d15d689a9ce2
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
ERROR: Exception:
Traceback (most recent call last):
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 164, in exc_logging_wrapper
    status = run_func(*args)
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 205, in wrapper
    return func(self, options, args)
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/commands/install.py", line 338, in run
    requirement_set = resolver.resolve(
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 73, in resolve
    collected = self.factory.collect_root_requirements(root_reqs)
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 468, in collect_root_requirements
    req = self._make_requirement_from_install_req(
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 430, in _make_requirement_from_install_req
    cand = self._make_candidate_from_link(
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 201, in _make_candidate_from_link
    self._link_candidate_cache[link] = LinkCandidate(
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 281, in __init__
    super().__init__(
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 156, in __init__
    self.dist = self._prepare()
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 225, in _prepare
    dist = self._prepare_distribution()
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 292, in _prepare_distribution
    return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 482, in prepare_linked_requirement
    return self._prepare_linked_requirement(req, parallel_builds)
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 546, in _prepare_linked_requirement
    dist = _get_prepared_distribution(
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 58, in _get_prepared_distribution
    abstract_dist.prepare_distribution_metadata(finder, build_isolation)
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/distributions/sdist.py", line 47, in prepare_distribution_metadata
    self._install_build_reqs(finder)
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/distributions/sdist.py", line 106, in _install_build_reqs
    build_reqs = self._get_build_requires_wheel()
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_internal/distributions/sdist.py", line 83, in _get_build_requires_wheel
    return backend.get_requires_for_build_wheel()
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_vendor/pep517/wrappers.py", line 172, in get_requires_for_build_wheel
    return self._call_hook('get_requires_for_build_wheel', {
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_vendor/pep517/wrappers.py", line 332, in _call_hook
    raise BackendUnavailable(data.get('traceback', ''))
pip._vendor.pep517.wrappers.BackendUnavailable: Traceback (most recent call last):
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 89, in _build_backend
    obj = import_module(mod_path)
  File "/usr/local/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 972, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/private/var/folders/k6/2dzjkwsd1mj8m52xz1b2swym0000gp/T/pip-build-env-hp7wsna4/overlay/lib/python3.9/site-packages/poetry/core/masonry/__init__.py", line 10, in <module>
    from .builder import Builder
  File "/private/var/folders/k6/2dzjkwsd1mj8m52xz1b2swym0000gp/T/pip-build-env-hp7wsna4/overlay/lib/python3.9/site-packages/poetry/core/masonry/builder.py", line 7, in <module>
    from .builders.sdist import SdistBuilder
  File "/private/var/folders/k6/2dzjkwsd1mj8m52xz1b2swym0000gp/T/pip-build-env-hp7wsna4/overlay/lib/python3.9/site-packages/poetry/core/masonry/builders/__init__.py", line 2, in <module>
    from .wheel import WheelBuilder
  File "/private/var/folders/k6/2dzjkwsd1mj8m52xz1b2swym0000gp/T/pip-build-env-hp7wsna4/overlay/lib/python3.9/site-packages/poetry/core/masonry/builders/wheel.py", line 23, in <module>
    from packaging.tags import sys_tags
  File "/private/var/folders/k6/2dzjkwsd1mj8m52xz1b2swym0000gp/T/pip-build-env-hp7wsna4/overlay/lib/python3.9/site-packages/poetry/core/_vendor/packaging/tags.py", line 7, in <module>
    import distutils.util
  File "/Users/alkasm/prog/setuptools-60.0.0-install-error/venv/lib/python3.9/site-packages/_distutils_hack/__init__.py", line 92, in create_module
    return importlib.import_module('setuptools._distutils')
  File "/usr/local/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
ModuleNotFoundError: No module named 'setuptools'
```
</details>

### Code of Conduct

- [X] I agree to follow the PSF Code of Conduct

**Comment by layday** [comment]
> Alright, so the problem is that setuptools is using import hooks which is not something that pip's build isolation isolates against.

As I explained 7 days ago.  I understand how asottile might have missed since he's blocked me (apparently that's a thing we do in the packaging community -  just block people we disagree with), not sure how everybody else has.

**Comment by jaraco** [comment]
> How does that [_distutils_hack] work `@jaraco`?

I meant to respond to this sooner, then got distracted and it fell off my radar. The hack is pretty simple. A top-level module [_distutils_hack](https://github.com/pypa/setuptools/tree/main/_distutils_hack) is installed and loaded by a [distutils-precedence.pth](https://github.com/pypa/setuptools/blob/dcdeedbed77711549f41e1f46f186c94ef9c4657/setup.py#L55-L58) file. All that `_distutils_hack.ensure_shim` does is add a `MetaPathFinder` to find distutils inside Setuptools.

An open question to the pip team (@pradyunsg) is what are [these lines](https://github.com/pypa/pip/blob/5cf98408f48a0ef91d61aea56485a7a83f6bbfa8/src/pip/_internal/build_env.py#L107-L108) intended to accomplish? Looking at the [whole block](https://github.com/pypa/pip/blob/5cf98408f48a0ef91d61aea56485a7a83f6bbfa8/src/pip/_internal/build_env.py#L104-L117), it looks like the sitecustomize explicitly adds `system_site` dirs in order to remove them, which has the side effect of handling .pth files on that path.

It's not obvious to me that there's anything a `sitecustomize` file can do to prevent the execution of the `.pth` hooks in system site packages, because [the site packages are unconditionally added before site-customize is called](https://github.com/python/cpython/blob/77195cd44b2506cda88a3cfc98918526068b1d46/Lib/site.py#L603-L609). Pip will probably need to run with `python -S` if it wishes to suppress .pth handling in system site-packages.

I want to explore whether the MetaPathFinder should be more lenient when its accompanying module cannot be found. #2962 captures a related problem where the hook exists but picks up an unrelated setuptools module.

**Comment by jaraco** [comment]
Released as 60.1.1. Please test and let me know if there are outstanding issues.

**Comment by rnickle** [comment]
> Released as 60.1.1. Please test and let me know if there are outstanding issues.

I ran into this issue in one of my CI flows with poetry 1.1.12 and setuptools 60.1.0

I tried to fix it by simply applying 60.1.1, but that did not fix it.   I had to additionally downgrade  poetry to 1.1.11.

**Comment by jaraco** [comment]
I was worried that there was some race between fixes between poetry 1.1.12 and setuptools 60.1.1, but I'm unable to replicate the issue even with poetry 1.1.12:

```
draft $ venv/bin/pip install -U setuptools
Requirement already satisfied: setuptools in ./venv/lib/python3.10/site-packages (60.1.0.post20211228)
Collecting setuptools
  Downloading setuptools-60.1.1-py3-none-any.whl (952 kB)
     |████████████████████████████████| 952 kB 3.0 MB/s 
Installing collected packages: setuptools
  Attempting uninstall: setuptools
    Found existing installation: setuptools 60.1.0.post20211228
    Uninstalling setuptools-60.1.0.post20211228:
      Successfully uninstalled setuptools-60.1.0.post20211228
Successfully installed setuptools-60.1.1
WARNING: You are using pip version 21.2.3; however, version 21.3.1 is available.
You should consider upgrading via the '/Users/jaraco/draft/venv/bin/python -m pip install --upgrade pip' command.
draft $ venv/bin/pip install -v --no-clean --force-reinstall --no-binary poetry --no-deps poetry
Using pip 21.2.3 from /Users/jaraco/draft/venv/lib/python3.10/site-packages/pip (python 3.10)
Collecting poetry
  Using cached poetry-1.1.12.tar.gz (135 kB)
  Running command /Users/jaraco/draft/venv/bin/python /private/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-standalone-pip-pys6gzws/__env_pip__.zip/pip install --ignore-installed --no-user --prefix /private/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-build-env-cp7ylkg8/overlay --no-warn-script-location --no-binary poetry --only-binary :none: -i https://pypi.org/simple -- 'poetry-core>=1.0.0'
  Collecting poetry-core>=1.0.0
    Using cached poetry_core-1.0.7-py2.py3-none-any.whl (424 kB)
  Installing collected packages: poetry-core
  Successfully installed poetry-core-1.0.7
  WARNING: You are using pip version 21.2.3; however, version 21.3.1 is available.
  You should consider upgrading via the '/Users/jaraco/draft/venv/bin/python -m pip install --upgrade pip' command.
  Installing build dependencies ... done
  Running command /Users/jaraco/draft/venv/bin/python /Users/jaraco/draft/venv/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py get_requires_for_build_wheel /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmpxnq7m2n8
  Getting requirements to build wheel ... done
    Running command /Users/jaraco/draft/venv/bin/python /Users/jaraco/draft/venv/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmpz59by_ip
    Preparing wheel metadata ... done
Building wheels for collected packages: poetry
  Running command /Users/jaraco/draft/venv/bin/python /Users/jaraco/draft/venv/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmppexg7ssv
  Building wheel for poetry (PEP 517) ... done
  Created wheel for poetry: filename=poetry-1.1.12-py2.py3-none-any.whl size=175030 sha256=3d4dc5d4c7171fb251b32e49a98cecbe2795d5e681ecee2297e2b2d2ccd671fc
  Stored in directory: /Users/jaraco/Library/Caches/pip/wheels/ef/8f/20/5291aa91f96e4d6f87a0de0eb1169d19ee1c6a7f61eeb9068f
Successfully built poetry
Installing collected packages: poetry
  Attempting uninstall: poetry
    Found existing installation: poetry 1.1.12
    Uninstalling poetry-1.1.12:
      Removing file or directory /Users/jaraco/draft/venv/bin/poetry
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry-1.1.12.dist-info/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/__init__.py
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/__main__.py
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/__pycache__/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/__version__.py
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/_vendor/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/config/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/console/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/exceptions.py
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/factory.py
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/inspection/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/installation/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/io/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/json/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/layouts/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/locations.py
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/masonry/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/mixology/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/packages/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/poetry.py
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/publishing/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/puzzle/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/repositories/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/utils/
      Removing file or directory /Users/jaraco/draft/venv/lib/python3.10/site-packages/poetry/version/
      Successfully uninstalled poetry-1.1.12
  changing mode of /Users/jaraco/draft/venv/bin/poetry to 755
Successfully installed poetry-1.1.12
WARNING: You are using pip version 21.2.3; however, version 21.3.1 is available.
You should consider upgrading via the '/Users/jaraco/draft/venv/bin/python -m pip install --upgrade pip' command.
```
